qualtran/bloqs/mod_arithmetic/mod_multiplication.py

```diff
@@ -265,3 +273,463 @@ def _modmul_symb() -> CModMulK:
+        reduced: whether a final modular reduction was applied.
+
+    References:
+        [Performance Analysis of a Repetition Cat Code Architecture: Computing 256-bit Elliptic Curve Logarithm in 9 Hours with 126 133 Cat Qubits](https://arxiv.org/abs/2302.06639)
+            Appendix C4.
+
+        [How to compute a 256-bit elliptic curve private key with only 50 million Toffoli gates](https://arxiv.org/abs/2306.08585)
+            page 8.
+    """
+    bitsize: 'SymbolicInt'
+    window_size: 'SymbolicInt'
+    mod: 'SymbolicInt'
+    uncompute: bool = False
+
+    def __attrs_post_init__(self):
+        if isinstance(self.mod, int):
+            assert self.mod > 1 and self.mod % 2 == 1  # Must be an odd integer greater than 1.
+
+        if isinstance(self.mod, int) and isinstance(self.bitsize, int):
+            assert 2 * self.mod - 1 < 2**self.bitsize, f'bitsize={self.bitsize} is too small'
+
+        if isinstance(self.window_size, int) and isinstance(self.bitsize, int):
+            assert self.bitsize % self.window_size == 0
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        num_windows = (
+            self.bitsize + self.window_size - 1
+        ) // self.window_size  # = ceil(self.bitsize/self.window_size)
+        side = Side.LEFT if self.uncompute else Side.RIGHT
+        return Signature(
+            [
+                Register('x', QMontgomeryUInt(self.bitsize)),
+                Register('y', QMontgomeryUInt(self.bitsize)),
+                Register('target', QMontgomeryUInt(self.bitsize), side=side),
+                Register(
+                    'qrom_indices', QMontgomeryUInt(num_windows * self.window_size), side=side
+                ),
+                Register('reduced', QBit(), side=side),
+            ]
+        )
+
+    def adjoint(self) -> 'DirtyOutOfPlaceMontgomeryModMul':
+        return attrs.evolve(self, uncompute=self.uncompute ^ True)
+
+    @cached_property
+    def _inversion_data(self) -> np.typing.NDArray:
+        inv_mod = pow(self.mod, 2 ** (self.window_size - 1) - 1, 2**self.window_size)
+        N = 2**self.window_size
+        data = (-np.arange(N) * inv_mod) % N
+        data *= self.mod
+        return data
+
+    def _classical_action_window(
+        self,
+        x: 'ClassicalValT',
+        y: 'ClassicalValT',
+        target: 'ClassicalValT',
+        qrom_indices: 'ClassicalValT',
+    ):
+        if is_symbolic(self.bitsize) or is_symbolic(self.window_size) or is_symbolic(self.mod):
+            raise ValueError(f'classical action is not supported for {self}')
+        for i in range(self.window_size):
+            if (x >> i) & 1:
+                target += y << i
+        m = target & (2**self.window_size - 1)
+        Tm = self._inversion_data[m]
+        target += Tm
+        target >>= self.window_size
+        qrom_indices = (qrom_indices << self.window_size) | m
+        return target, qrom_indices
```

Comment:
this is now duplicate code, right? See my other comment about how we can make the classical action more readable / useful

Reply:
no .. this is part of the classical action function for DirtyOutOfPlaceMontgomeryModMul

Reply:
this code is nearly identical to `SingleWindowModMul.on_classical_vals`. Can you explain the distinction (if any) between the two, preferably in code comments

Reply:
the main difference is that  SingleWindowModMul.on_classical_vals operates on bitarrays. this methods operates on ints. 

added a comment.